### PR TITLE
Handle ISingleValueCaseClauseOperation and IDefaultCaseClauseOperation in control flow graph

### LINKED ISSUE #23795: ISingleValueCaseClauseOperation is missing a Value for some scenarios, which means that a portion of user code is not represented in the IOperation tree
Here is an existing IOperation test:
```
        <Fact, WorkItem(17607, "https://github.com/dotnet/roslyn/issues/17607")>
        Public Sub InvalidSwitchStatementCaseLabel()
            Dim source = <![CDATA[
Class Program
    Public Shared Sub Main(args As String())
        Dim x = New Program()
        Select Case x.ToString()'BIND:"Select Case x.ToString()"
            Case x
                Exit Select
        End Select
    End Sub
End Class]]>.Value

            Dim expectedOperationTree = <![CDATA[
ISwitchOperation (1 cases) (OperationKind.Switch, Type: null, IsInvalid) (Syntax: 'Select Case ... End Select')
  Switch expression: 
    IInvocationOperation (virtual Function System.Object.ToString() As System.String) (OperationKind.Invocation, Type: System.String) (Syntax: 'x.ToString()')
      Instance Receiver: 
        ILocalReferenceOperation: x (OperationKind.LocalReference, Type: Program) (Syntax: 'x')
      Arguments(0)
  Sections:
      ISwitchCaseOperation (1 case clauses, 1 statements) (OperationKind.SwitchCase, Type: null, IsInvalid) (Syntax: 'Case x ... Exit Select')
          Clauses:
              ISingleValueCaseClauseOperation (CaseKind.SingleValue) (OperationKind.CaseClause, Type: null, IsInvalid) (Syntax: 'x')
                Value: 
                  null
          Body:
              IBlockOperation (1 statements) (OperationKind.Block, Type: null, IsInvalid, IsImplicit) (Syntax: 'Case x ... Exit Select')
                IBranchOperation (BranchKind.Break, Label: exit) (OperationKind.Branch, Type: null) (Syntax: 'Exit Select')
]]>.Value

            Dim expectedDiagnostics = <![CDATA[
BC30311: Value of type 'Program' cannot be converted to 'String'.
            Case x
                 ~
]]>.Value

            VerifyOperationTreeAndDiagnosticsForTest(Of SelectBlockSyntax)(source, expectedOperationTree, expectedDiagnostics)
        End Sub
```

Note that reference to local ```x``` in ```Case x``` is not present in the tree:
```
              ISingleValueCaseClauseOperation (CaseKind.SingleValue) (OperationKind.CaseClause, Type: null, IsInvalid) (Syntax: 'x')
                Value: 
                  null
```

Success scenarios are affected too, SelectCase_ImplicitOperator unit-test for example.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Also fixes #23795.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: >/// [](start = 36, length = 3)

Extra slashes.

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >TryCast [](start = 36, length = 7)

Can you point me at the test where this cast fails?

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: > Can you point me at the test where this cast fails?

I don't have a test like this. Do you see any problem with the logic?

Comment (333fred) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: My only problem is that where before we had a DirectCast, we only have a TryCast now. I didn't see any tests with changed Syntax based on this, so I'm wondering why the change.

---
In reply to: [184238556](https://github.com/dotnet/roslyn/pull/26395#discussion_r184238556) [](ancestors = 184238556)

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: I think it is much better to avoid making decisions based on syntax nodes associated with bound nodes. But it is Ok to try to peek a "better" syntax for an IOperation node when we can, and it doesn't make sense to crash when we cannot.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >!section.Clauses.Any(c => c.Label != null) [](start = 33, length = 42)

This simplifies to `section.Clauses.All(c => c.Label == null)`, which is also significantly less mindbending to read and comprehend.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.cs]: Not seeing equivalent added tests for VB.

**Comment by 333fred** [comment]
Done review (Iteration 2)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > TODO what? It would be good to have a list

This is coming in my next PRs

---
In reply to: [184241555](https://github.com/dotnet/roslyn/pull/26395#discussion_r184241555) [](ancestors = 184241555)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Why are these left in?

I am planning to handle these in several next PRs, they are not handled at the moment.

---
In reply to: [184241599](https://github.com/dotnet/roslyn/pull/26395#discussion_r184241599) [](ancestors = 184241599)

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: The ```// PROTOTYPE(dataflow): TODO.``` comment is about them.

---
In reply to: [184263471](https://github.com/dotnet/roslyn/pull/26395#discussion_r184263471) [](ancestors = 184263471,184241599)

File: src/Compilers/Core/Portable/Operations/IPlaceholderOperation.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/IPlaceholderOperation.cs]: > Missing documentation

This is an internal API. It is self-explanatory

---
In reply to: [184241643](https://github.com/dotnet/roslyn/pull/26395#discussion_r184241643) [](ancestors = 184241643)

File: src/Compilers/Core/Portable/Operations/IPlaceholderOperation.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/IPlaceholderOperation.cs]: > Convention so far for *Kinds has been to put them in a different file.

I believe it is much better to keep this enum in the same file where IPlaceholderOperation is defined

---
In reply to: [184241682](https://github.com/dotnet/roslyn/pull/26395#discussion_r184241682) [](ancestors = 184241682)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.cs]: > Not seeing equivalent added tests for VB.

You are correct. I didn't need to add extra tests for this in VB because I didn't have to synthesize nodes for similar VB scenarios in ControlFlowGraphBuilder.

---
In reply to: [184243761](https://github.com/dotnet/roslyn/pull/26395#discussion_r184243761) [](ancestors = 184243761)

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: > My only problem is that where before we had a DirectCast, we only have a TryCast now. I didn't see any tests with changed Syntax based on this, so I'm wondering why the change.

The old code was checking syntax Kind before doing DirectCast, so the cast was safe. The new code doesn't check the syntax kind first.

---
In reply to: [184238987](https://github.com/dotnet/roslyn/pull/26395#discussion_r184238987) [](ancestors = 184238987,184238556)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need second review.

**Comment by AlekseyTs** [comment]
@mavasani, @dotnet/analyzer-ioperation Please review

**Comment by AlekseyTs** [comment]
@mavasani, @dotnet/analyzer-ioperation Please review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review, need second review.

**Review by 333fred — APPROVED** [review]
LGTM with followup in future PR.

**Review by mavasani — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: `If(expr Is Nothing, OperationFactory.NullOperation, New Lazy(Of IOperation)(Function() Create(expr)))` is used here while `New Lazy(Of IOperation)(Function() Create(expr))` is used in most other cases. We should be consistent, ideally with a `CreateLazyOperation(exprOpt As BoundExpression)` helper that returns `OperationFactory.NullOperation` if `exprOpt Is Nothing`.

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: Sure, will switch to ```New Lazy(Of IOperation)(Function() Create(boundCaseBlock.CaseStatement.ConditionOpt))```  

---
In reply to: [184538767](https://github.com/dotnet/roslyn/pull/26395#discussion_r184538767) [](ancestors = 184538767)

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs
Comment (cston) [inline src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: Consider using `Assert.False(explicitNodeMap.ContainsKey(descendant.Syntax), $"Duplicate ...")` rather than `try/catch`.

Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: > Consider using Assert.False(explicitNodeMap.ContainsKey(descendant.Syntax), $"Duplicate ...") rather than try/catch.

I need to add key to the dictionary as well. Killing two birds with one stone.

---
In reply to: [184550508](https://github.com/dotnet/roslyn/pull/26395#discussion_r184550508) [](ancestors = 184550508)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Consider getting the `SemanticModel` and `Compilation` once, in the .ctor, and storing in `readonly` fields.

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Consider getting the SemanticModel and Compilation once, in the .ctor, and storing in readonly fields.

Will add a prototype comment for that 

---
In reply to: [184554290](https://github.com/dotnet/roslyn/pull/26395#discussion_r184554290) [](ancestors = 184554290)

**Review by cston — APPROVED** [review]

**Pull request merged**
